### PR TITLE
Enable System.Net.Sockets.Tests on netfx 

### PR DESCRIPTION
**Pull request by stephentoub**
- Adds a netfx test config
- Adds a simple SocketTaskExtensions to the test project to be used in the netfx config
- Fixes or skips tests on netfx configuration

cc: @davidsh, @safern
Closes https://github.com/dotnet/corefx/issues/18152
Contributes to https://github.com/dotnet/corefx/issues/17690

### CODE REVIEW

File: src/System.Net.Sockets/tests/FunctionalTests/ArgumentValidationTests.cs [hunk 1]
```diff
@@ -503,6 +503,7 @@ public void Select_LargeList_Throws_ArgumentOutOfRange()
             Assert.Throws<ArgumentOutOfRangeException>(() => Socket.Select(null, null, largeList, -1));
         }
 
+        [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework, "Bug in AcceptAsync that dereferences null SAEA argument")]
```
Comment (davidsh) [inline src/System.Net.Sockets/tests/FunctionalTests/ArgumentValidationTests.cs]: We need an open GitHub issue to track investigating/fixing tests/product code.  You don't need to open an different issue for each test you disable. You can do what I've done already and have an issue for each library.  Please then list that issue number in the SkipOnTargetFramework attribute.

File: src/System.Net.Sockets/tests/FunctionalTests/Configurations.props [hunk 2]
```diff
@@ -3,6 +3,7 @@
   <PropertyGroup>
     <BuildConfigurations>
       netstandard;
+      netfx;
     </BuildConfigurations>
   </PropertyGroup>
 </Project>
```
Comment (davidsh) [inline src/System.Net.Sockets/tests/FunctionalTests/Configurations.props]: nit: missing newline at end of file.

**Comment by safern** [comment]
I think the problem of SocketTaskExtensions not being available in netfx will be fixed on #18246 according to the conversation on that PR. So I don't know if we should only disable te tests and wait for that PR to go in.

**Comment by stephentoub** [comment]
> I think the problem of SocketTaskExtensions not being available in netfx will be fixed on #18246 according to the conversation on that PR

Ah, great, I didn't realize that was already happening.  We can wait for that to go in then, and I'll delete the file from here.

File: src/System.Net.Sockets/tests/FunctionalTests/Configurations.props (hunk 2 above)
Comment (stephentoub) [inline src/System.Net.Sockets/tests/FunctionalTests/Configurations.props]: That was already the case prior to this change.

File: src/System.Net.Sockets/tests/FunctionalTests/ArgumentValidationTests.cs (hunk 1 above)
Comment (stephentoub) [inline src/System.Net.Sockets/tests/FunctionalTests/ArgumentValidationTests.cs]: > We need an open GitHub issue to track investigating/fixing tests/product code

We're using GitHub issues to track known/understood bugs in desktop?

**Comment by davidsh** [comment]
>Ah, great, I didn't realize that was already happening. We can wait for that to go in then, and I'll delete the file from here.

I don't understand exactly how we will use that new Desktop Support package here other than to maybe just bring in the SocketTaskExtensions.* source file?

**Comment by stephentoub** [comment]
> I don't understand exactly how we will use that new Desktop Support package 

It looks like that package includes an implementation of SocketTaskExtensions, so presumably the tests just include a reference to that package.

File: src/System.Net.Sockets/tests/FunctionalTests/ArgumentValidationTests.cs (hunk 1 above)
Comment (davidsh) [inline src/System.Net.Sockets/tests/FunctionalTests/ArgumentValidationTests.cs]: >We're using GitHub issues to track known/understood bugs in desktop?

No. But does that mean that the test is being disabled due to a bug in NETFX and not a test issue?

Comment (stephentoub) [inline src/System.Net.Sockets/tests/FunctionalTests/ArgumentValidationTests.cs]: > But does that mean that the test is being disabled due to a bug in NETFX and not a test issue?

Yes, e.g. in this particular one you commented on, netfx is missing argument validation on the argument and just dereferences it immediately, so with netfx you get a NullReferenceException, whereas with core you get an ArgumentNullException.

Comment (davidsh) [inline src/System.Net.Sockets/tests/FunctionalTests/ArgumentValidationTests.cs]: Ok

**Comment by safern** [comment]
> It looks like that package includes an implementation of SocketTaskExtensions

Yes the it includes an implementation of SocketTaskExtensions, see: https://github.com/dotnet/corefx/pull/18246/files#diff-19774c11d07be2307c4d7a15947b86c5

And see how this file will be included only when building against netfx configuration: https://github.com/ericstj/corefx/blob/1ae9e8156c198b8d2f1b060c38c7b218ed333baa/src/System.Net.Sockets/src/System.Net.Sockets.csproj#L31-L33

I'm not sure if we will need to reference the package or it will do its magic itself as a netfx configuration is being added to the src and ref contract.

**Comment by stephentoub** [comment]
Cool.  Then I'll just wait for that.

**Comment by ericstj** [comment]
I restored [that implementation](https://github.com/dotnet/corefx/pull/18246/files#diff-19774c11d07be2307c4d7a15947b86c5) from 1.1 since @stephentoub made a lot of changes in master that made it not work for desktop.  @stephentoub if you want to refine that implementation after my PR please do so.

**Comment by stephentoub** [comment]
Thanks, @ericstj.  The 1.1 implementation should generally be fine; most of the changes I made since then really only work well with access to Sockets internals.

**Pull request merged**